go-controller/pkg/ovn/controller/egressservice_zone/egressservice_zone_service.go

```diff
@@ -127,41 +128,71 @@ func (c *Controller) onServiceDelete(obj interface{}) {
 	c.egressServiceQueue.Add(key)
 }
 
-// Returns all of the non-host endpoints for the given service grouped by IPv4/IPv6.
-func (c *Controller) allEndpointsFor(svc *corev1.Service) (sets.Set[string], sets.Set[string], error) {
+// Returns cluster-networked endpoints for the given service grouped by IPv4/IPv6.
+// When IC is disabled v[4|6]LocalEndpoints contains all service endpoints and v[4|6]RemoteEndpoints is not set
+// When IC is enabled v[4|6]LocalEndpoints contains endpoints hosted in the local zone and
+// v[4|6]RemoteEndpoints contains endpoints hosted in remote zones
+func (c *Controller) allEndpointsFor(svc *corev1.Service) (
+	v4LocalEndpoints sets.Set[string],
+	v6LocalEndpoints sets.Set[string],
+	v4RemoteEndpoints sets.Set[string],
+	v6RemoteEndpoints sets.Set[string],
+	err error) {
 	// Get the endpoint slices associated to the Service
 	esLabelSelector := labels.Set(map[string]string{
 		discovery.LabelServiceName: svc.Name,
 	}).AsSelectorPreValidated()
 
 	endpointSlices, err := c.endpointSliceLister.EndpointSlices(svc.Namespace).List(esLabelSelector)
 	if err != nil {
-		return nil, nil, err
+		return
 	}
 
-	v4Endpoints := sets.New[string]()
-	v6Endpoints := sets.New[string]()
+	v4LocalEndpoints = sets.Set[string]{}
+	v6LocalEndpoints = sets.Set[string]{}
+	v4RemoteEndpoints = sets.Set[string]{}
+	v6RemoteEndpoints = sets.Set[string]{}
+
 	for _, eps := range endpointSlices {
 		if eps.AddressType == discovery.AddressTypeFQDN {
 			continue
 		}
 
-		epsToInsert := v4Endpoints
+		localEndpoints := v4LocalEndpoints
+		remoteEndpoints := v4RemoteEndpoints
 		if eps.AddressType == discovery.AddressTypeIPv6 {
-			epsToInsert = v6Endpoints
+			localEndpoints = v6LocalEndpoints
+			remoteEndpoints = v6RemoteEndpoints
 		}
 
 		for _, ep := range eps.Endpoints {
+			if ep.NodeName == nil {
+				// ignore endpoints without a node
+				continue
+			}
+			isEpLocal := true
+			if config.OVNKubernetesFeature.EnableInterconnect {
+				var zoneKnown bool
+				zoneKnown, isEpLocal = c.nodeZoneState(*ep.NodeName)
+				if !zoneKnown {
+					klog.Errorf("Failed to get the zone for %v", ep)
+					continue
+				}
+			}
 			for _, ip := range ep.Addresses {
 				ipStr := utilnet.ParseIPSloppy(ip).String()
 				if !services.IsHostEndpoint(ipStr) {
-					epsToInsert.Insert(ipStr)
+					if isEpLocal {
+						localEndpoints.Insert(ipStr)
+					} else {
+						remoteEndpoints.Insert(ipStr)
+					}
+
 				}
 			}
 		}
 	}
-
-	return v4Endpoints, v6Endpoints, nil
+	return
```

Comment:
wdyt about changing this to a regular return https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters? I'm not a big fan of it personally :sweat_smile:

Reply:
I think what I'm talking about relates to the 
```
Naked returns are okay if the function is a handful of lines. Once it's a medium sized function, be explicit with your return values
```
:sweat_smile:

Reply:
I think the named returns are still less confusing considering the amount of values we return. 
I think a potential improvement that doesn't have to happen in this pr is to return two slices of structs that would contain all the info we require.